misc/bstr.c

```diff
@@ -454,3 +454,27 @@ struct bstr bstr_get_ext(struct bstr s)
         return (struct bstr){NULL, 0};
     return bstr_splice(s, dotpos + 1, s.len);
 }
+
+struct bstr bstr_decode_hex(void *talloc_ctx, struct bstr hex)
+{
+    char *out = talloc_array(talloc_ctx, char, hex.len);
+    int pos = 0;
+
+    // Read out all hex chars as individual nibbles
+    for (int i = 0; i < hex.len; i++) {
+        unsigned char in = hex.start[i];
+        if (in >= '0' && in <= '9')
+            out[pos++] = in - '0';
+        if (in >= 'a' && in <= 'f')
+            out[pos++] = in - 'a' + 10;
+        if (in >= 'A' && in <= 'F')
+            out[pos++] = in - 'A' + 10;
+    }
+
+    // Go through the output and recombine all of the nibbles in-place
+    int len = pos / 2;
+    for (int i = 0; i < len; i++)
+        out[i] = (out[2*i+0] << 4) | out[2*i+1];
```

Comment:
@wm4 How would you parse '0\n3' in-place? Seems like “parse first nibble, skip all non-hex, parse next nibble, combine both before writing” is more complexity than doing it this way. It would also require an extra conditional to make sure there even *exists* a next nibble (and you don't start reading out of bounds), because you're unsure whether the number of nibbles is even or odd at the time of parsing the first one.